### PR TITLE
Add bit information to generalized FP generator

### PR DESCRIPTION
**Pull request by greglandrum**
more detail to come when this is actually ready to review

### CODE REVIEW

File: Code/GraphMol/Fingerprints/FingerprintGenerator.h [hunk 1]
```diff
@@ -15,31 +15,58 @@
...
 
 namespace RDKit {
 class ROMol;
 
 struct RDKIT_FINGERPRINTS_EXPORT AdditionalOutput {
-  // will review this structure once more fingerprint types are implemented
+  using atomToBitsType = std::vector<std::vector<std::uint64_t>>;
+  using bitInfoMapType =
+      std::map<std::uint64_t,
+               std::vector<std::pair<std::uint32_t, std::uint32_t>>>;
```
Comment (ricrogz) [inline Code/GraphMol/Fingerprints/FingerprintGenerator.h:29]: Probably not, but just to make sure, should these be updated to `uint64_t` too ?

File: Code/GraphMol/Fingerprints/RDKitFPGenerator.cpp [hunk 2]
```diff
@@ -90,21 +90,32 @@ RDKitFPArguments<OutputType>::RDKitFPArguments(
...
-    const bool                           // hashResults
-) const {
-  // todo set additional outputs
+    FingerprintArguments<OutputType> *,                    // arguments
+    const std::vector<std::uint32_t> *,                    // atomInvariants
+    const std::vector<std::uint32_t> *,                    // bondInvariants
+    const AdditionalOutput *additionalOutput, const bool,  // hashResults
+    const std::uint64_t fpSize) const {
+  if (additionalOutput) {
+    OutputType bit_id = fpSize ? (d_bitId % fpSize) : d_bitId;
```
Comment (ricrogz) [inline Code/GraphMol/Fingerprints/RDKitFPGenerator.cpp:99]: The ternary operator is not too readable... I generally prefer to use and `if` statement. Maybe:
```suggestion
    OutputType bit_id = d_bitId;
    if (fpSize) {
        bit_id %= fpSize;
    }
```

File: Code/GraphMol/Fingerprints/TopologicalTorsionGenerator.cpp [hunk 3]
```diff
@@ -44,17 +44,29 @@ OutputType TopologicalTorsionAtomEnv<OutputType>::getBitId(
...
     const std::vector<std::uint32_t> *,  // atomInvariants
     const std::vector<std::uint32_t> *,  // bondInvariants
-    const AdditionalOutput *,            // additionalOutput
-    const bool                           // hashResults
-) const {
+    const AdditionalOutput *additionalOutput,
+    const bool,  // hashResults
+    const std::uint64_t fpSize) const {
+  if (additionalOutput) {
+    auto bitId = fpSize ? (d_bitId % fpSize) : d_bitId;
```
Comment (ricrogz) [inline Code/GraphMol/Fingerprints/TopologicalTorsionGenerator.cpp:51]: Again, for readability, maybe:
```suggestion
    auto bitId = d_bitId;
    if (fpSize) {
        bitId %= fpSize;
    }
```

**Review by ricrogz — APPROVED** [review]
I'm not sure I'm a too good reviewer for this, I don't know how these fingerprints work, but what we do here looks sensible to me. I'm just adding a few minor suggestions about formatting/readability.

File: Code/GraphMol/Fingerprints/FingerprintGenerator.h (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/Fingerprints/FingerprintGenerator.h:29]: unfortunately not

File: Code/GraphMol/Fingerprints/catch_tests.cpp [hunk 4]
```diff
@@ -144,4 +142,610 @@ TEST_CASE("pattern fingerprints for MolBundles", "[patternfp]") {
...
+                                                    2,
+                                                }}},
+                                               {1308, {{0}, {1}}},
+                                               {1339, {{2}}},
+                                               {1728, {{1, 2}}},
+                                               {1813,
+                                                {{
+                                                    0,
+                                                    1,
+                                                }}}};
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/Fingerprints/catch_tests.cpp:311]: nice. I didn't know about `// clang-format off`. That really helps

**Comment by greglandrum** [comment]
> I'm not sure I'm a too good reviewer for this, I don't know how these fingerprints work, but what we do here looks sensible to me. I'm just adding a few minor suggestions about formatting/readability.

Thanks for doing the review!

File: Code/GraphMol/Fingerprints/catch_tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/Fingerprints/catch_tests.cpp]: Yeah, it's good to have when "formatting horrors" like these happen :)

Although simple as it is, I keep forgetting the exact syntax, and mixing it up with yapf's, so in the end I always end up googling it again and again...

**Comment by ricrogz** [comment]
> > I'm not sure I'm a too good reviewer for this, I don't know how these fingerprints work, but what we do here looks sensible to me. I'm just adding a few minor suggestions about formatting/readability.
> 
> Thanks for doing the review!

No problem!

**Pull request merged**